docs/advanced/fields.rst

```diff
@@ -0,0 +1,206 @@
+            return self.response.css(".name").get()
+
+        def price(self):
+            return self.response.css(".price").get()
+
+        def to_item(self) -> dict:
+            return {
+                'name': self.name(),
+                'price': self.price()
+            }
+
+This approach has 2 main advantages:
+
+1. Often the code looks cleaner this way, it's easier to follow.
+2. The resulting page object becomes more flexible and reusable:
+   if not all data extracted in the ``to_item()`` method is needed,
+   user can call extraction methods for individual attributes. It's
+   more efficient than running ``to_item()`` and only using some of the
+   result.
+
+However, writing and maintaining ``to_item()`` method can get tedious,
+especially if there is a lot of extraction methods.
+
+web_poet.fields
+---------------
+
+To aid writing Page Objects in this style, ``web-poet`` provides
+a few utilities:
+
+* :func:`@web_poet.field <web_poet.fields.field>` decorator,
+* :func:`web_poet.item_from_fields <web_poet.fields.item_from_fields>`
+  and :func:`web_poet.item_from_fields_sync <web_poet.fields.item_from_fields_sync>`
+  functions.
+
+We can rewrite the example like this:
+
+.. code-block:: python
+
+    import attrs
+    from web_poet import ItemPage, HttpResponse, field, item_from_fields_sync
+
+
+    @attrs.define
+    class MyPage(ItemPage):
+        response: HttpResponse
+
+        @field
+        def name(self):
+            return self.response.css(".name").get()
+
+        @field
+        def price(self):
+            return self.response.css(".price").get()
+
+        def to_item(self) -> dict:
+            return item_from_fields_sync(self)
+
+
+As you can guess, :func:`~.item_from_fields_sync` uses all the methods annotated
+with :func:`@field <web_poet.fields.field>` decorator, and creates a dict
+with the result, where keys are method names, and values are results of
+calling these methods.
+
+Asynchronous extraction methods
+-------------------------------
+
+``async def`` extraction methods are also supported, as well as a mix of
+sync and async methods - use :func:`~.item_from_fields` for this.
+For example, you might send :ref:`advanced-requests` to extract some
+of the attributes:
+
+.. code-block:: python
+
+    import attrs
+    from web_poet import ItemPage, HttpResponse, HttpClient, field, item_from_fields
+
+
+    @attrs.define
+    class MyPage(ItemPage):
+        response: HttpResponse
+        http_client: HttpClient
+
+        @field
+        def name(self):
+            return self.response.css(".name").get()
+
+        @field
+        async def price(self):
+            resp = self.http_client.get("...")
+            return resp.json()['price']
+
+        async def to_item(self) -> dict:
+            return item_from_fields(self)
+
+Because :func:`~.item_from_fields` supports both sync and async extraction
+methods, it's recommended to use it over :func:`~.item_from_fields_sync`, even
```

Comment:
What are the specific scenarios that we're only allowed to use the sync function instead of the async? (given that the latter supports both cases)

For context, I'm wondering if we should not expose the sync function to make the choice easier for users.

Reply:
The only justification for the `_sync` version I can think of would be a slight performance improvement. However, if it is significant and we can not reduce the performance gap between both methods, I think the documentation needs to indicate so, and otherwise, I agree that it may be best not to provide `_sync` in the first place.

Reply:
I'm fine with making it private. Main motivation to include it was to allow using field decorators with page objects where to_item method is not async.

Reply:
Right, that is a valid point.

I have no strong opinion on providing both vs forcing users to use `async def to_item`, although I am slightly in favor of the later.